kernel/src/fs/named_pipe.rs

```diff
@@ -1,50 +1,230 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use super::{
-    file_handle::FileLike,
-    pipe::{self, PipeReader, PipeWriter},
-    utils::{AccessMode, Metadata},
-};
+use ostd::sync::{Rcu, WaitQueue};
+
+use super::pipe::{self, PipeReader, PipeWriter};
 use crate::{
     events::IoEvents,
+    fs::{
+        inode_handle::FileIo,
+        utils::{AccessMode, StatusFlags},
+    },
     prelude::*,
     process::signal::{PollHandle, Pollable},
 };
 
-pub struct NamedPipe {
+/// A handle representing the reader end of a named pipe.
+struct ReaderHandle {
     reader: Arc<PipeReader>,
+}
+
+impl Drop for ReaderHandle {
+    fn drop(&mut self) {
+        self.reader.state.peer_shutdown();
+    }
+}
+
+/// A handle representing the writer end of a named pipe.
+struct WriterHandle {
     writer: Arc<PipeWriter>,
 }
 
-impl NamedPipe {
-    pub fn new() -> Result<Self> {
-        let (reader, writer) = pipe::new_pair()?;
+impl Drop for WriterHandle {
+    fn drop(&mut self) {
+        self.writer.state.shutdown();
+    }
+}
+
+/// A handle for an opened named pipe that encapsulates different access modes.
+///
+/// This enum represents the different ways a named pipe can be opened:
+/// - Read-only: Only reading operations are allowed
+/// - Write-only: Only writing operations are allowed  
+/// - Read-write: Both reading and writing operations are allowed
+enum NamedPipeHandle {
+    Reader(Arc<ReaderHandle>),
+    Writer(Arc<WriterHandle>),
+    Both {
+        reader: Arc<ReaderHandle>,
+        writer: Arc<WriterHandle>,
+    },
+}
```

Comment:
Do we readlly need such an enum? Why not just let `NamedPipe::open` return `Arc<ReaderHandle>`, `Arc<WriterHandle>` and `Arc<ReaderWriterHandle>` according to the access mode? By this way, you can avoid an Arc wrapper for the read-only/write-only case.